frontend/src/metabase/parameters/utils/dashboards.ts

```diff
@@ -109,17 +109,41 @@ function getMappings(dashcards: DashboardCard[]): ExtendedMapping[] {
 
     return extendedParameterMappings;
   });
+
+  return extendedParameterMappings;
+}
+
+function getMappingsByParameterId(dashcards: DashboardCard[]) {
+  const mappings = getMappings(dashcards);
+  const mappingsByParameterId = new Map<string, Map<symbol, ExtendedMapping>>();
+
+  for (const mapping of mappings) {
+    if (!mappingsByParameterId.has(mapping.parameter_id)) {
+      mappingsByParameterId.set(mapping.parameter_id, new Map());
+    }
+
+    const targetKey = Symbol.for(mapping.target.toString());
+    if (!mappingsByParameterId.get(mapping.parameter_id)?.has(targetKey)) {
+      mappingsByParameterId.get(mapping.parameter_id)?.set(targetKey, mapping);
+    }
+  }
+
+  return mappingsByParameterId;
 }
 
 export function getDashboardUiParameters(
   dashboard: Dashboard,
   metadata: Metadata,
 ): UiParameter[] {
   const { parameters, dashcards } = dashboard;
-  const mappings = getMappings(dashcards as DashboardCard[]);
+  const mappings = getMappingsByParameterId(dashcards as DashboardCard[]);
   const uiParameters: UiParameter[] = (parameters || []).map(parameter => {
     if (isFieldFilterParameter(parameter)) {
-      return buildFieldFilterUiParameter(parameter, mappings, metadata);
+      return buildFieldFilterUiParameter(
+        parameter,
+        [...(mappings.get(parameter.id)?.values() ?? [])],
```

Comment:
nit: there are many things going on here, and it feels a bit hard to parse
might be worth unpacking it into several lines